lib/connection/connection_config.js

```diff
@@ -756,6 +773,28 @@ function ConnectionConfig(options, validateCredentials, qaMode, clientInfo)
 
   // custom agent class, test only
   this.agentClass = options.agentClass;
+
+  function getOrigin(reqUrl) {
+    const parsed = url.parse(reqUrl);
+    const protocol = parsed.protocol || 'http:';
+    const port = parsed.port || (protocol === 'http:' ? '80' : '443');
+    return `${protocol}//${parsed.hostname}:${port}`;
+  }
+
+  const agentCache = new Map();
+
+  this.getCachedAgent = function (reqUrl)
+  {
+    const key = getOrigin(reqUrl);
+    return agentCache.get(key);
+  };
+
+  this.setCachedAgent = function (reqUrl, agent)
+  {
+    const key = getOrigin(reqUrl);
+    Logger.getInstance().debug(`Caching agent for origin '${key}'`);
+    agentCache.set(key, agent);
+  };
```

Comment:
For `keepAlive` to work correctly, we need to cache Agent instances somewhere. Each unique origin should have its own Agent (which may optionally keep the socket alive if `keepAlive` is enabled).

I figured the best place is to add this cache is the connection config object itself. (I avoided using a singleton for this because it complicates the integration tests).